bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FolderWatcherHandler.java

```diff
@@ -0,0 +1,263 @@
+    private List<String> previousFtpListing = new ArrayList<>();
+
+    public FolderWatcherHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(FolderWatcherConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+        if (config.connectionTimeout <= 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Connection timeout can't be negative");
+            return;
+        }
+        if (config.ftpPort < 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "FTP port can't be negative");
+            return;
+        }
+        if (config.ftpDir == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "FTP directory can't be empty");
+            return;
+        }
+
+        currentFtpListingFile = new File(ConfigConstants.getUserDataFolder() + File.separator + "FolderWatcher"
+                + File.separator + thing.getUID().getAsString().replace(':', '_') + ".data");
+        try {
+            previousFtpListing = WatcherCommon.initStorage(currentFtpListingFile);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            logger.debug("Can't write file {}, error message {}", currentFtpListingFile, e.getMessage());
+            return;
+        }
+
+        if (config.pollInterval > 0) {
+            initJob = scheduler.scheduleWithFixedDelay(this::connectionKeepAlive, 0, config.pollInterval,
+                    TimeUnit.SECONDS);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Polling interval can't be null or negative");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (executionJob != null) {
+            executionJob.cancel(true);
+        }
+        if (initJob != null) {
+            initJob.cancel(true);
+        }
+        if (ftp.isConnected()) {
+            try {
+                ftp.logout();
+                ftp.disconnect();
+            } catch (IOException e) {
+                logger.debug("Error terminating FTP connection: ", e);
+            }
+        }
+    }
+
+    private List<String> listDirectory(@Nullable FTPClient ftpClient, String parentDir, String currentDir, int level,
+            boolean recursive) throws IOException {
+        Instant dateNow = Instant.now();
+        List<String> dirList = new ArrayList<>();
+        final String dirToList = parentDir + ((!currentDir.equals("")) ? ("/" + currentDir) : "");
+
+        Optional.ofNullable(ftpClient.listFiles(dirToList)).map(Arrays::stream).orElseGet(Stream::empty)
+                .forEach(file -> {
```

Comment:
Lambdas aren't really helping the readability here. Can you just write this with a regular loop?

Reply:
@cpmeister thanks for your inputs, using of lambdas was driven by previous code review. Originally I had normal loop.

Reply:
Yeah, the code here isn't very readable nor is it performant or easily maintainable; please switch it back to a regular loop.

Reply:
Yeah, I disagree with J-N-K's implementation here. Lambdas and optionals can solve many problems, but they shouldn't be used to solve ALL problems.